var/spack/repos/builtin/packages/papi/package.py

```diff
@@ -64,6 +68,18 @@ def configure_args(self):
                           self.spec.variants)
         if variants:
             options.append('--with-components={0}'.format(' '.join(variants)))
+
+        # The logic above enables specific components
+        # The logic below explicitly disable variants when specified
+        if '~static' in spec:
+            options.append('--with-static-lib=no')
+
+        if '~shared' in spec:
+            options.append('--with-shared-lib=no')
+
+        if '~shlib_tools' in spec:
+            options.append('--without-shlib')
```

Comment:
Can you enable these in the if clause and explicitly disable them in the else clause?

Reply:
@adamjstewart  Its already in there, though its different than most other packages. Papi supports building components by pasing in a list of the enabled components and that is done here:
```python
variants = filter(lambda x: self.spec.variants[x].value is True,
                          self.spec.variants)
        if variants:
            options.append('--with-components={0}'.format(' '.join(variants)))
```
Simply by adding the new variants, they are explicitly enabled. I added the logic this comment points to in order to explicitly disable them.

Reply:
Interesting. Do we need to explicitly disable any other variants?

Reply:
Admittedly I'm not an expert on building this package, but I think all of them can be explicitly disabled so I think it just boils down to preference. I didn't make that change since I didn't want to potentially break something, but if you want I can convert every variant to being explicitly disabled. 

It may be possible to disable packages using a similar scheme to how they are being enabled, though I have not tried that. Personally, I am not a huge fan of this method of enabling/disabling variants since all it takes is for one option to not be supported in this list method for the whole package to break.

Reply:
> Personally, I am not a huge fan of this method of enabling/disabling variants since all it takes is for one option to not be supported in this list method for the whole package to break.

I understand that (although I think there's an Autotools flag to ignore unknown flags). But the downside of not explicitly disabling things is that builds systems like Autotools and CMake like to toggle default values based on whether or not dependencies are found on the system. This leads to non-deterministic builds, and installations that link to system libraries. That makes binary installation a nightmare, and can make it difficult to debug build failures. You can also end up with different/incompatible versions of the same library in the DAG if the system libraries get sucked in. So there's pros and cons of being explicit for sure. Just wanted to explain where I'm coming from.

Reply:
No I totally agree, that makes sense. I meant I'm not a fan of passing in an entire list of every variant like the package is currently doing. Right now the package just iterates through every variant and, if enabled, appends it to the `--with-components=` list. 

While it might be less elegant (more lines), I'm suggesting replacing that with a explicit toggle on/off for each variant. While the appending variants to that list option might work now, its possible that a future variant that we want wont play nicely with this list method. 

Either way I support having explicit enable/disable toggles for each variant.

Reply:
Ah sorry, I completely misunderstood. Yes, that makes sense. @G-Ragghianti is the package maintainer, so maybe he has some opinions on the use of `--with-components`.

Reply:
Since these new variants are not papi "components", you'll need to change the lambda on line 70 to include only the variant names that are papi components.

Reply:
I agree that if we are going to include support for (e.g.) --with-shlib-tools, that spack should pass explicit "yes" or "no" rather than just leave it off instead of "no".  This will avoid the confusing behaviour of a user asking for "papi~shlib_tools" but still getting the shlib-tools (same for static and shared options).

Reply:
Ok. I will make those changes on the new variants. Should we just convert every variant to being handled like that or would you rather keep the lambda for the other variants that are Papi components?

Reply:
In the "filer()" you would replace self.spec.variants with `['example', 'infiniband', 'powercap', 'rapl', 'lmsensors', 'sde']`